File: python/tests/test_identity.py
"""Tests for IdentityRouter — multi-identity adapter routing."""

from __future__ import annotations

import asyncio
from typing import AsyncIterator

import pytest

from unified_channel.adapter import ChannelAdapter
from unified_channel.identity import IdentityRouter
from unified_channel.types import (
    ChannelStatus,
    ContentType,
    Identity,
    MessageContent,
    OutboundMessage,
    UnifiedMessage,
)


def _make_msg(
    channel: str = "telegram",
    text: str = "hello",
    msg_id: str = "m1",
    chat_id: str = "chat1",
) -> UnifiedMessage:
    return UnifiedMessage(
        id=msg_id,
        channel=channel,
        sender=Identity(id="user1", username="user1"),
        content=MessageContent(type=ContentType.TEXT, text=text),
        chat_id=chat_id,
    )


class FakeAdapter(ChannelAdapter):
    """In-memory adapter for testing."""

    def __init__(self, channel_id: str = "telegram", messages: list[UnifiedMessage] | None = None):
        self.channel_id = channel_id
        self.sent: list[OutboundMessage] = []
        self._messages = messages or []
        self.connected = False
        self.disconnected = False

    async def connect(self) -> None:
        self.connected = True

    async def disconnect(self) -> None:
        self.disconnected = True

    async def receive(self) -> AsyncIterator[UnifiedMessage]:
        for msg in self._messages:
            yield msg

    async def send(self, msg: OutboundMessage) -> str | None:
        self.sent.append(msg)
        return f"sent-{len(self.sent)}"

    async def get_status(self) -> ChannelStatus:
        return ChannelStatus(connected=self.connected, channel=self.channel_id)


class FailingAdapter(FakeAdapter):
    """Adapter whose get_status raises."""

    async def get_status(self) -> ChannelStatus:
        raise ConnectionError("status check failed")


# ── Registration ─────────────────────────────────────────────────────


def test_register_and_list():
    router = IdentityRouter()
    adapter = FakeAdapter("telegram")
    router.register("telegram:personal", adapter)
    assert router.get_identities() == ["telegram:personal"]


def test_register_duplicate_raises():
    router = IdentityRouter()
    router.register("telegram:personal", FakeAdapter("telegram"))
    with pytest.raises(ValueError, match="already registered"):
        router.register("telegram:personal", FakeAdapter("telegram"))


def test_register_invalid_format_raises():
    router = IdentityRouter()
    with pytest.raises(ValueError, match="invalid identity_id"):
        router.register("no-colon", FakeAdapter("telegram"))


def test_register_invalid_chars_raises():
    router = IdentityRouter()
    with pytest.raises(ValueError, match="invalid identity_id"):
        router.register("telegram:my-personal", FakeAdapter("telegram"))


def test_unregister():
    router = IdentityRouter()
    router.register("telegram:work", FakeAdapter("telegram"))
    router.unregister("telegram:work")
    assert router.get_identities() == []


def test_unregister_unknown_raises():
    router = IdentityRouter()
    with pytest.raises(KeyError, match="not registered"):
        router.unregister("telegram:nope")


@pytest.mark.asyncio
async def test_unregister_clears_default():
    router = IdentityRouter()
    router.register("telegram:main", FakeAdapter("telegram"))
    router.set_default("telegram", "telegram:main")
    router.unregister("telegram:main")
    # Default should be cleared
    with pytest.raises(KeyError, match="no default"):
        await router.send_default("telegram", OutboundMessage(chat_id="c", text="hi"))


# ── Send ─────────────────────────────────────────────────────────────


@pytest.mark.asyncio
async def test_send_to_specific_identity():
    router = IdentityRouter()
    adapter = FakeAdapter("telegram")
    router.register("telegram:personal", adapter)

    msg = OutboundMessage(chat_id="123", text="hello")
    result = await router.send("telegram:personal", msg)

    assert result == "sent-1"
    assert len(adapter.sent) == 1
    assert adapter.sent[0].text == "hello"


@pytest.mark.asyncio
async def test_send_unknown_identity_raises():
    router = IdentityRouter()
    with pytest.raises(KeyError, match="not registered"):
        await router.send("telegram:ghost", OutboundMessage(chat_id="c", text="hi"))


@pytest.mark.asyncio
async def test_send_default():
    router = IdentityRouter()
    personal = FakeAdapter("telegram")
    work = FakeAdapter("telegram")
    router.register("telegram:personal", personal)
    router.register("telegram:work", work)
    router.set_default("telegram", "telegram:work")

    msg = OutboundMessage(chat_id="123", text="via default")
    await router.send_default("telegram", msg)

    assert len(work.sent) == 1
    assert len(personal.sent) == 0
    assert work.sent[0].text == "via default"


@pytest.mark.asyncio
async def test_send_default_no_default_raises():
    router = IdentityRouter()
    router.register("telegram:personal", FakeAdapter("telegram"))
    with pytest.raises(KeyError, match="no default"):
        await router.send_default("telegram", OutboundMessage(chat_id="c", text="hi"))


def test_set_default_unknown_identity_raises():
    router = IdentityRouter()
    with pytest.raises(KeyError, match="not registered"):
        router.set_default("telegram", "telegram:nope")


def test_set_default_wrong_channel_raises():
    router = IdentityRouter()
    router.register("slack:team", FakeAdapter("slack"))
    with pytest.raises(ValueError, match="does not belong"):
        router.set_default("telegram", "slack:team")


# ── Receive ──────────────────────────────────────────────────────────


@pytest.mark.asyncio
async def test_receive_all_from_multiple_identities():
    msg1 = _make_msg(channel="telegram", text="from personal", msg_id="m1")
    msg2 = _make_msg(channel="telegram", text="from work", msg_id="m2")

    router = IdentityRouter()
    router.register("telegram:personal", FakeAdapter("telegram", messages=[msg1]))
    router.register("telegram:work", FakeAdapter("telegram", messages=[msg2]))

    received: list[tuple[str, UnifiedMessage]] = []
    async for identity_id, msg in router.receive_all():
        received.append((identity_id, msg))

    assert len(received) == 2
    ids = {iid for iid, _ in received}
    assert ids == {"telegram:personal", "telegram:work"}
    texts = {m.content.text for _, m in received}
    assert texts == {"from personal", "from work"}


@pytest.mark.asyncio
async def test_receive_all_empty():
    router = IdentityRouter()
    received = []
    async for item in router.receive_all():
        received.append(item)
    assert received == []


# ── get_identities filtering ─────────────────────────────────────────


def test_get_identities_filter_by_channel():
    router = IdentityRouter()
    router.register("telegram:a", FakeAdapter("telegram"))
    router.register("telegram:b", FakeAdapter("telegram"))
    router.register("slack:main", FakeAdapter("slack"))

    assert sorted(router.get_identities("telegram")) == ["telegram:a", "telegram:b"]
    assert router.get_identities("slack") == ["slack:main"]
    assert router.get_identities("discord") == []


def test_get_identities_all():
    router = IdentityRouter()
    router.register("telegram:a", FakeAdapter("telegram"))
    router.register("slack:b", FakeAdapter("slack"))
    assert sorted(router.get_identities()) == ["slack:b", "telegram:a"]


# ── connect_all / disconnect_all ─────────────────────────────────────


@pytest.mark.asyncio
async def test_connect_all():
    router = IdentityRouter()
    a1 = FakeAdapter("telegram")
    a2 = FakeAdapter("slack")
    router.register("telegram:main", a1)
    router.register("slack:main", a2)

    await router.connect_all()

    assert a1.connected
    assert a2.connected


@pytest.mark.asyncio
async def test_disconnect_all():
    router = IdentityRouter()
    a1 = FakeAdapter("telegram")
    a2 = FakeAdapter("slack")
    router.register("telegram:main", a1)
    router.register("slack:main", a2)

    await router.disconnect_all()

    assert a1.disconnected
    assert a2.disconnected


# ── Status ───────────────────────────────────────────────────────────


@pytest.mark.asyncio
async def test_get_status_all():
    router = IdentityRouter()
    a1 = FakeAdapter("telegram")
    a1.connected = True
    a2 = FakeAdapter("slack")
    router.register("telegram:main", a1)
    router.register("slack:main", a2)

    statuses = await router.get_status_all()

    assert statuses["telegram:main"].connected is True
    assert statuses["slack:main"].connected is False


@pytest.mark.asyncio
async def test_get_status_all_handles_errors():
    router = IdentityRouter()
    router.register("telegram:broken", FailingAdapter("telegram"))

    statuses = await router.get_status_all()

    assert statuses["telegram:broken"].connected is False
    assert "status check failed" in (statuses["telegram:broken"].error or "")


# ── Multiple same-channel identities ────────────────────────────────


@pytest.mark.asyncio
async def test_multiple_same_channel_send_independently():
    router = IdentityRouter()
    personal = FakeAdapter("telegram")
    work = FakeAdapter("telegram")
    router.register("telegram:personal", personal)
    router.register("telegram:work", work)

    await router.send("telegram:personal", OutboundMessage(chat_id="a", text="p"))
    await router.send("telegram:work", OutboundMessage(chat_id="b", text="w"))

    assert len(personal.sent) == 1
    assert personal.sent[0].text == "p"
    assert len(work.sent) == 1
    assert work.sent[0].text == "w"


# ── Chaining API ────────────────────────────────────────────────────


def test_register_returns_self():
    router = IdentityRouter()
    result = router.register("telegram:a", FakeAdapter("telegram"))
    assert result is router


def test_fluent_chaining():
    router = (
        IdentityRouter()
        .register("telegram:a", FakeAdapter("telegram"))
        .register("slack:b", FakeAdapter("slack"))
        .set_default("telegram", "telegram:a")
    )
    assert len(router.get_identities()) == 2

File: rust/tests/test_identity.rs
//! Tests for IdentityRouter: register, send, defaults, validation.

use async_trait::async_trait;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;
use tokio::sync::Mutex;

use unified_channel::adapter::{ChannelAdapter, MessageHandler};
use unified_channel::error::Result;
use unified_channel::identity::IdentityRouter;
use unified_channel::types::*;

/// Mock adapter for identity router tests.
struct MockAdapter {
    id: String,
    send_count: Arc<AtomicUsize>,
    last_text: Arc<Mutex<String>>,
}

impl MockAdapter {
    fn new(id: &str) -> Self {
        Self {
            id: id.to_string(),
            send_count: Arc::new(AtomicUsize::new(0)),
            last_text: Arc::new(Mutex::new(String::new())),
        }
    }
}

#[async_trait]
impl ChannelAdapter for MockAdapter {
    fn channel_id(&self) -> &str {
        &self.id
    }
    async fn connect(&mut self) -> Result<()> {
        Ok(())
    }
    async fn disconnect(&mut self) -> Result<()> {
        Ok(())
    }
    fn on_message(&mut self, _handler: MessageHandler) {}
    async fn send(&self, msg: OutboundMessage) -> Result<Option<String>> {
        self.send_count.fetch_add(1, Ordering::SeqCst);
        let mut last = self.last_text.lock().await;
        *last = msg.text.clone();
        Ok(Some("sent_id".into()))
    }
    async fn get_status(&self) -> ChannelStatus {
        ChannelStatus::connected(&self.id)
    }
}

#[tokio::test]
async fn test_register_and_send() {
    let adapter = MockAdapter::new("telegram");
    let count = Arc::clone(&adapter.send_count);

    let mut router = IdentityRouter::new();
    router
        .register("telegram:personal", Box::new(adapter))
        .unwrap();

    let msg = OutboundMessage::text("chat1", "hello");
    let result = router.send("telegram:personal", msg).await;
    assert!(result.is_ok());
    assert_eq!(count.load(Ordering::SeqCst), 1);
}

#[tokio::test]
async fn test_register_invalid_id() {
    let mut router = IdentityRouter::new();
    let result = router.register("invalid-format", Box::new(MockAdapter::new("x")));
    assert!(result.is_err());
    let err = result.err().unwrap();
    assert!(err.to_string().contains("invalid identity_id"));
}

#[tokio::test]
async fn test_register_duplicate() {
    let mut router = IdentityRouter::new();
    router
        .register("telegram:one", Box::new(MockAdapter::new("telegram")))
        .unwrap();

    let result = router.register("telegram:one", Box::new(MockAdapter::new("telegram")));
    assert!(result.is_err());
    let err = result.err().unwrap();
    assert!(err.to_string().contains("already registered"));
}

#[tokio::test]
async fn test_send_unknown_identity() {
    let router = IdentityRouter::new();
    let msg = OutboundMessage::text("chat1", "hello");
    let result = router.send("telegram:unknown", msg).await;
    assert!(result.is_err());
}

#[tokio::test]
async fn test_set_default_and_send() {
    let adapter = MockAdapter::new("telegram");
    let count = Arc::clone(&adapter.send_count);

    let mut router = IdentityRouter::new();
    router
        .register("telegram:personal", Box::new(adapter))
        .unwrap();
    router.set_default("telegram", "telegram:personal").unwrap();

    let msg = OutboundMessage::text("chat1", "via default");
    let result = router.send_default("telegram", msg).await;
    assert!(result.is_ok());
    assert_eq!(count.load(Ordering::SeqCst), 1);
}

#[tokio::test]
async fn test_set_default_unregistered_identity() {
    let mut router = IdentityRouter::new();
    let result = router.set_default("telegram", "telegram:missing");
    assert!(result.is_err());
}

#[tokio::test]
async fn test_set_default_channel_mismatch() {
    let mut router = IdentityRouter::new();
    router
        .register("discord:main", Box::new(MockAdapter::new("discord")))
        .unwrap();

    let result = router.set_default("telegram", "discord:main");
    assert!(result.is_err());
    let err = result.err().unwrap();
    assert!(err.to_string().contains("does not belong"));
}

#[tokio::test]
async fn test_send_default_no_default_set() {
    let mut router = IdentityRouter::new();
    router
        .register("telegram:one", Box::new(MockAdapter::new("telegram")))
        .unwrap();

    let msg = OutboundMessage::text("chat1", "hello");
    let result = router.send_default("telegram", msg).await;
    assert!(result.is_err());
    assert!(result.unwrap_err().to_string().contains("no default"));
}

#[tokio::test]
async fn test_unregister() {
    let mut router = IdentityRouter::new();
    router
        .register("telegram:one", Box::new(MockAdapter::new("telegram")))
        .unwrap();
    assert_eq!(router.len(), 1);

    router.unregister("telegram:one").unwrap();
    assert_eq!(router.len(), 0);
    assert!(router.is_empty());
}

#[tokio::test]
async fn test_unregister_cleans_default() {
    let mut router = IdentityRouter::new();
    router
        .register("telegram:one", Box::new(MockAdapter::new("telegram")))
        .unwrap();
    router.set_default("telegram", "telegram:one").unwrap();
    router.unregister("telegram:one").unwrap();

    let msg = OutboundMessage::text("chat1", "hello");
    let result = router.send_default("telegram", msg).await;
    assert!(result.is_err()); // default was cleaned up
}

#[tokio::test]
async fn test_get_identities() {
    let mut router = IdentityRouter::new();
    router
        .register("telegram:a", Box::new(MockAdapter::new("telegram")))
        .unwrap();
    router
        .register("telegram:b", Box::new(MockAdapter::new("telegram")))
        .unwrap();
    router
        .register("discord:c", Box::new(MockAdapter::new("discord")))
        .unwrap();

    let all = router.get_identities(None);
    assert_eq!(all.len(), 3);

    let tg = router.get_identities(Some("telegram"));
    assert_eq!(tg.len(), 2);

    let dc = router.get_identities(Some("discord"));
    assert_eq!(dc.len(), 1);

    let slack = router.get_identities(Some("slack"));
    assert_eq!(slack.len(), 0);
}

#[tokio::test]
async fn test_get_status_all() {
    let mut router = IdentityRouter::new();
    router
        .register("telegram:a", Box::new(MockAdapter::new("telegram")))
        .unwrap();
    router
        .register("discord:b", Box::new(MockAdapter::new("discord")))
        .unwrap();

    let statuses = router.get_status_all().await;
    assert_eq!(statuses.len(), 2);
    // Both should show connected (our mock returns connected)
    assert!(statuses["telegram:a"].connected);
    assert!(statuses["discord:b"].connected);
}
